src/gennylib/include/gennylib/quiesce.hpp

```diff
@@ -0,0 +1,138 @@
+            mongocxx::client client(mongocxx::uri(desc.primaryUri));
+            auto admin = client["admin"];
+
+            // Assert that all nodes are in reasonable states.
+            auto replSetRes = admin.run_command(make_document(kvp("replSetGetStatus", 1)));
+            auto members = replSetRes.view()["members"];
+            if (members && members.type() == bsoncxx::type::k_array) {
+                bsoncxx::array::view members_view = members.get_array();
+                for (auto member : members_view) {
+                    std::string state(member["stateStr"].get_utf8().value);
+                    if (state != "PRIMARY" && state != "SECONDARY" && state != "ARBITER") {
+                        std::string name(member["name"].get_utf8().value);
+                        BOOST_LOG_TRIVIAL(error) << "Cannot wait oplog, replset member "
+                            << name << " is " << state;
+                        success_acc = false;
+                        return;
+                    }
+                }
+            }
+
+            // Do flush
+            auto collection = admin["wait_oplog"];
+            mongocxx::write_concern wc;
+            wc.nodes(desc.nodes.size());
+            wc.journal(true);
+            mongocxx::options::insert opts;
+            opts.write_concern(wc);
+            auto insertRes = collection.insert_one(make_document(kvp("x", "flush")), opts);
+        
+            success_acc = success_acc && insertRes && insertRes->result().inserted_count() == 1;
+        }
+        bool success_acc = true;
+    };
+
+    WaitOplogVisitor waitVisitor;
+    topology.accept(waitVisitor);
+
+    return waitVisitor.success_acc;
+}
+
+void doFsync(Topology& topology) {
+    class DoFsyncVisitor : public TopologyVisitor {
+        void visitMongodDescription(const MongodDescription& desc) {
+            using bsoncxx::builder::basic::kvp;
+            using bsoncxx::builder::basic::make_document;
+
+            mongocxx::client client(mongocxx::uri(desc.mongodUri));
+            auto admin = client["admin"];
+            admin.run_command(make_document(kvp("fsync", 1)));
+        }
+    };
+    DoFsyncVisitor v;
+    topology.accept(v);
+}
+
+bool checkForDroppedCollections(mongocxx::database db) {
+    using bsoncxx::builder::basic::kvp;
+    using bsoncxx::builder::basic::make_document;
+
+    auto res = db.run_command(make_document(kvp("serverStatus", 1)));
+    auto idents = res.view()["storageEngine"]["dropPendingIdents"];
+
+    return idents.get_int64() != 0;
+}
+
+bool checkForDroppedCollectionsTestDB(mongocxx::pool::entry& client, std::string dbName) {
+    auto db = client->database(dbName);
+    int retries = 0;
+    while (checkForDroppedCollections(db) && retries < DROPPED_COLLECTION_RETRIES) {
+        BOOST_LOG_TRIVIAL(debug) << "Sleeping 1 second while waiting for collection to finish dropping.";
+        retries++;
+        std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
1. I'm super skeptical of how this will play with `PhaseLoop`. I'm holding my breath for now but note to self to come back here.
2. separate thought: do we want to ensure there aren't concurrent invocations of this?

Reply:
Added a locking mechanism to `quiesce.cpp` for the latter